File: src/modules/module_1/1.5_memory.py
from langchain.tools import ToolRuntime, tool
from langchain.agents import create_agent, AgentState
from langchain.chat_models import init_chat_model

model= init_chat_model(model='gpt-5-nano')

class CustomState(AgentState):
    user_name: str

@tool
def greet_user(runtime: ToolRuntime[CustomState],) -> str:
    """Greet the user by name."""
    user_name = runtime.state.get("user_name", None)
    return f"The user name is: {user_name}!"

personal_chef_prompt = """
    You are a personal chef assistant. 
    Your task is to provide personalized meal recommendations based on user 
    Ingredients and preferences.
    Greet the user personally by his name.
"""

# Putting it all together
memory = create_agent(
    model=model,
    system_prompt=personal_chef_prompt,
    tools=[greet_user],
    state_schema=CustomState,
)

File: src/modules/module_1/1.5_memory/1.5_memory.ts
import { createAgent, tool, initChatModel, type ToolRuntime } from "langchain";
import { StateSchema } from "@langchain/langgraph";
import z from "zod";

const model = await initChatModel("gpt-5-nano");

// CustomState extends the built-in agent state with a user_name field.
// StateSchema is the TypeScript equivalent of Python's AgentState subclass.
const CustomState = new StateSchema({
  user_name: z.string().optional(),
});

const greetUser = tool(
  // runtime.state gives access to the full agent state, including custom fields.
  async (_, runtime: ToolRuntime<typeof CustomState.State>) => {
    const userName = runtime.state.user_name ?? null;
    return `The user name is: ${userName}!`;
  },
  {
    name: "greet_user",
    description: "Greet the user by name.",
    schema: z.object({}),
  },
);

const personalChefPrompt = `
    You are a personal chef assistant.
    Your task is to provide personalized meal recommendations based on user
    Ingredients and preferences.
    Greet the user personally by his name.
`;

// Putting it all together
export const memory = createAgent({
  model,
  systemPrompt: personalChefPrompt,
  tools: [greetUser],
  stateSchema: CustomState,
});
